File: src/nyne_er_lab/live.py
"""Live profile resolution — resolve ad-hoc profiles against known identities."""

from __future__ import annotations

import hashlib
import urllib.request
import urllib.error
from dataclasses import dataclass
from datetime import datetime, timezone

from nyne_er_lab.cluster.resolver import ResolvedPair, generate_evidence_card
from nyne_er_lab.features.dataset import PairExample
from nyne_er_lab.features.extractor import PairFeatureExtractor
from nyne_er_lab.ingest import parse_raw_page
from nyne_er_lab.models.hybrid import TrainedHybridMatcher
from nyne_er_lab.schemas import (
    OrganizationClaim,
    ProfileRecord,
    RawProfilePage,
    TextSpan,
)


@dataclass
class LiveResult:
    """Result of resolving a live profile against the known corpus."""

    query_profile: ProfileRecord
    matches: list[ResolvedPair]
    best_match_profile: ProfileRecord | None
    best_match_identity_name: str | None


def _stable_id(text: str) -> str:
    return "live-" + hashlib.sha256(text.encode()).hexdigest()[:12]


def fetch_url(url: str, timeout: int = 10) -> str:
    """Fetch raw HTML from a URL. Returns empty string on failure."""
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "EntityResolutionLab/0.1"})
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            return resp.read().decode("utf-8", errors="replace")
    except (urllib.error.URLError, OSError, ValueError):
        return ""


def profile_from_url(
    url: str,
    source_type: str,
    display_name_hint: str | None = None,
) -> ProfileRecord | None:
    """Fetch a URL, parse its HTML into a ProfileRecord."""
    html = fetch_url(url)
    if not html:
        return None
    page_id = _stable_id(url)
    try:
        raw_page = RawProfilePage(
            page_id=page_id,
            source_type=source_type,
            url=url,
            display_name_hint=display_name_hint,
            fetched_at=datetime.now(timezone.utc),
            title=display_name_hint,
            html=html,
            raw_text=html[:5000],
        )
        return parse_raw_page(raw_page)
    except (ValueError, Exception):
        return None


def profile_from_text(
    display_name: str,
    bio: str,
    source_type: str = "personal_site",
    headline: str = "",
    organizations: list[str] | None = None,
    locations: list[str] | None = None,
    topics: list[str] | None = None,
    url: str = "https://example.com/profile",
) -> ProfileRecord:
    """Construct a ProfileRecord directly from user-supplied text fields."""
    orgs = organizations or []
    page_id = _stable_id(f"{display_name}-{bio[:50]}")
    bio_text = bio if bio.strip() else f"{display_name} professional profile."
    return ProfileRecord(
        profile_id=page_id,
        canonical_person_id=None,
        source_type=source_type,
        url=url,
        display_name=display_name,
        aliases=[],
        headline=headline or None,
        bio_text=bio_text,
        organizations=[OrganizationClaim(name=o) for o in orgs if o.strip()],
        education=[],
        locations=locations or [],
        outbound_links=[],
        topics=topics or [],
        timestamps=[],
        raw_text=f"{display_name} {headline} {bio_text}",
        supporting_spans=[TextSpan(field_name="bio_text", snippet=bio_text[:240])],
        metadata={},
    )


def resolve_live_profile(
    query: ProfileRecord,
    corpus_profiles: list[ProfileRecord],
    extractor: PairFeatureExtractor,
    matcher: TrainedHybridMatcher,
    identities,
) -> LiveResult:
    """Run the full pipeline on a single query profile against the corpus."""
    from nyne_er_lab.blocking.blocker import rule_reasons

    # Generate pairs between query and every corpus profile
    examples: list[PairExample] = []
    for corpus_profile in corpus_profiles:
        reasons = rule_reasons(query, corpus_profile)
        # Always generate a pair even without blocking reasons for live search
        features = extractor.featurize_pair(query, corpus_profile)
        examples.append(PairExample(
            left_profile_id=query.profile_id,
            right_profile_id=corpus_profile.profile_id,
            left_canonical_id="unknown",
            right_canonical_id=corpus_profile.canonical_person_id or "unknown",
            split="test",
            label=0,
            features=features,
            blocking_reasons=tuple(sorted(reasons)) if reasons else ("live_search",),
        ))

    if not examples:
        return LiveResult(query_profile=query, matches=[], best_match_profile=None, best_match_identity_name=None)

    scores, decisions = matcher.score_examples(examples, extractor)

    profile_lookup = {p.profile_id: p for p in corpus_profiles + [query]}
    resolved: list[ResolvedPair] = []
    for ex, score, decision in zip(examples, scores, decisions):
        left = profile_lookup[ex.left_profile_id]
        right = profile_lookup[ex.right_profile_id]
        evidence_card = generate_evidence_card(left, right, ex, score, decision)
        resolved.append(ResolvedPair(
            left_profile_id=ex.left_profile_id,
            right_profile_id=ex.right_profile_id,
            score=score,
            decision=decision,
            evidence_card=evidence_card,
        ))

    # Sort by score descending
    resolved.sort(key=lambda r: -r.score)

    # Find best match
    best = None
    best_identity = None
    for rp in resolved:
        if rp.decision == "match":
            best = profile_lookup.get(rp.right_profile_id)
            # Find identity containing this profile
            if identities:
                for identity in identities:
                    if rp.right_profile_id in identity.member_profile_ids:
                        best_identity = identity.canonical_name
                        break
            break

    return LiveResult(
        query_profile=query,
        matches=resolved,
        best_match_profile=best,
        best_match_identity_name=best_identity,
    )

File: crates/nyne-er-core/src/live.rs
use crate::blocking::rule_reasons;
use crate::cluster::{ResolvedPair, generate_evidence_card};
use crate::features::{PairExample, PairFeatureExtractor, SplitName};
use crate::models::TrainedHybridMatcher;
use crate::schemas::{OrganizationClaim, ProfileRecord, RawProfilePage, SourceType, TextSpan};
use anyhow::Result;
use chrono::Utc;
use reqwest::blocking::Client;
use reqwest::header::USER_AGENT;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;
use std::str::FromStr;
use url::Url;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TextProfileInput {
    pub display_name: String,
    pub bio: String,
    #[serde(default = "default_source_type")]
    pub source_type: String,
    #[serde(default)]
    pub headline: String,
    #[serde(default)]
    pub organizations: Vec<String>,
    #[serde(default)]
    pub locations: Vec<String>,
    #[serde(default)]
    pub topics: Vec<String>,
    #[serde(default = "default_profile_url")]
    pub url: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LiveResult {
    pub query_profile: ProfileRecord,
    pub matches: Vec<ResolvedPair>,
    pub best_match_profile: Option<ProfileRecord>,
    pub best_match_identity_name: Option<String>,
}

pub fn fetch_url(url: &str) -> String {
    let client = match Client::builder().build() {
        Ok(client) => client,
        Err(_) => return String::new(),
    };
    let response = match client
        .get(url)
        .header(USER_AGENT, "EntityResolutionLab/0.1")
        .send()
    {
        Ok(response) => response,
        Err(_) => return String::new(),
    };
    response.text().unwrap_or_default()
}

pub fn profile_from_url(
    url: &str,
    source_type: &str,
    display_name_hint: Option<&str>,
) -> Option<ProfileRecord> {
    let html = fetch_url(url);
    if html.is_empty() {
        return None;
    }
    let raw_page = RawProfilePage {
        page_id: stable_id(url),
        source_type: SourceType::from_str(source_type).ok()?,
        url: Url::parse(url).ok()?,
        canonical_person_id: None,
        display_name_hint: display_name_hint.map(str::to_string),
        fetched_at: Some(Utc::now()),
        title: display_name_hint.map(str::to_string),
        html: Some(html.clone()),
        raw_text: html.chars().take(5_000).collect(),
    }
    .validated()
    .ok()?;
    crate::ingest::parse_raw_page(&raw_page).ok()
}

pub fn profile_from_text(input: TextProfileInput) -> Result<ProfileRecord> {
    let source_type = SourceType::from_str(&input.source_type)?;
    let url = Url::parse(&input.url)?;
    let display_name = input.display_name;
    let headline = input.headline;
    let bio = input.bio;
    let bio_text = if bio.trim().is_empty() {
        format!("{display_name} professional profile.")
    } else {
        bio.trim().to_string()
    };
    ProfileRecord {
        profile_id: stable_id(&format!(
            "{}-{}",
            display_name,
            bio.chars().take(50).collect::<String>()
        )),
        canonical_person_id: None,
        source_type,
        url,
        display_name: display_name.clone(),
        aliases: Vec::new(),
        headline: (!headline.trim().is_empty()).then_some(headline.trim().to_string()),
        bio_text: bio_text.clone(),
        organizations: input
            .organizations
            .into_iter()
            .filter(|value| !value.trim().is_empty())
            .map(|value| OrganizationClaim {
                name: value,
                role: None,
                start_year: None,
                end_year: None,
            })
            .collect(),
        education: Vec::new(),
        locations: input.locations,
        outbound_links: Vec::new(),
        topics: input.topics,
        timestamps: Vec::new(),
        raw_text: format!("{} {} {}", display_name, headline, bio_text)
            .trim()
            .to_string(),
        supporting_spans: vec![TextSpan {
            field_name: "bio_text".to_string(),
            snippet: bio_text.chars().take(240).collect(),
            start_char: None,
            end_char: None,
        }],
        metadata: BTreeMap::new(),
    }
    .validated()
}

pub fn resolve_live_profile(
    query: &ProfileRecord,
    corpus_profiles: &[ProfileRecord],
    extractor: &PairFeatureExtractor,
    matcher: &TrainedHybridMatcher,
    identities: &[crate::schemas::CanonicalIdentity],
) -> Result<LiveResult> {
    let examples: Vec<PairExample> = corpus_profiles
        .iter()
        .map(|corpus_profile| PairExample {
            left_profile_id: query.profile_id.clone(),
            right_profile_id: corpus_profile.profile_id.clone(),
            left_canonical_id: "unknown".to_string(),
            right_canonical_id: corpus_profile
                .canonical_person_id
                .clone()
                .unwrap_or_else(|| "unknown".to_string()),
            split: SplitName::Test,
            label: 0,
            features: extractor.featurize_pair(query, corpus_profile),
            blocking_reasons: {
                let reasons = rule_reasons(query, corpus_profile);
                if reasons.is_empty() {
                    vec!["live_search".to_string()]
                } else {
                    reasons.into_iter().collect()
                }
            },
        })
        .collect();
    if examples.is_empty() {
        return Ok(LiveResult {
            query_profile: query.clone(),
            matches: Vec::new(),
            best_match_profile: None,
            best_match_identity_name: None,
        });
    }

    let (scores, decisions) = matcher.score_examples(&examples, extractor, None);
    let profile_lookup: BTreeMap<String, ProfileRecord> = corpus_profiles
        .iter()
        .cloned()
        .chain(std::iter::once(query.clone()))
        .map(|profile| (profile.profile_id.clone(), profile))
        .collect();

    let mut resolved = Vec::new();
    for ((example, score), decision) in examples.iter().zip(scores.iter()).zip(decisions.iter()) {
        let left = profile_lookup.get(&example.left_profile_id).expect("query exists");
        let right = profile_lookup.get(&example.right_profile_id).expect("candidate exists");
        resolved.push(ResolvedPair {
            left_profile_id: example.left_profile_id.clone(),
            right_profile_id: example.right_profile_id.clone(),
            score: *score,
            decision: decision.clone(),
            evidence_card: generate_evidence_card(left, right, example, *score, decision)?,
        });
    }
    resolved.sort_by(|left, right| {
        right
            .score
            .partial_cmp(&left.score)
            .unwrap_or(std::cmp::Ordering::Equal)
    });

    let mut best_match_profile = None;
    let mut best_match_identity_name = None;
    for pair in &resolved {
        if pair.decision != "match" {
            continue;
        }
        best_match_profile = profile_lookup.get(&pair.right_profile_id).cloned();
        best_match_identity_name = identities
            .iter()
            .find(|identity| identity.member_profile_ids.contains(&pair.right_profile_id))
            .map(|identity| identity.canonical_name.clone());
        break;
    }

    Ok(LiveResult {
        query_profile: query.clone(),
        matches: resolved,
        best_match_profile,
        best_match_identity_name,
    })
}

fn stable_id(text: &str) -> String {
    let digest = Sha256::digest(text.as_bytes());
    format!("live-{:x}", digest)[..17].to_string()
}

fn default_source_type() -> String {
    "personal_site".to_string()
}

fn default_profile_url() -> String {
    "https://example.com/profile".to_string()
}
